src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,733 @@
+    //
+    // RETURN: true on success, false on failure.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_system_err on POSIX open(2) failure
+    bool open_nx(const char *path) noexcept;
+    bool open(const char *path);
+
+    bool open_nx(const char *path, const int flags) noexcept;
+    bool open(const char *path, const int flags);
+
+    bool open_nx(const char *path, const int flags, const mode_t mode) noexcept;
+    bool open(const char *path, const int flags, const mode_t mode);
+
+    // Flush the buffer and close the current file descriptor.
+    //
+    // RETURN: true on success, false on failure cases.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time)
+    // dio::iostream_system_err on POSIX I/O failure
+    bool close_nx() noexcept;
+    bool close();
+
+    // RETURN: current bits/vars in stream based on io_states enum.
+    int current_state() noexcept;
+
+    // RETURN: true when there is no set bits/vars otherwise false.
+    bool good() noexcept;
+
+    // RETURN: true when buffer_failbit is set otherwise false.
+    bool buffer_failure() noexcept;
+
+    // RETURN: 0 or POSIX errno based on io_error value.
+    int io_failure() noexcept;
+
+    // Resets failure bits/vars to false (or 0 about io_error).
+    void clear() noexcept;
+
+    // Flush buffer of stream.
+    //
+    // RETURN: the size of buffer on full-flushed buffer, -1 on failure or not full-flushed buffer.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time)
+    // dio::iostream_system_err on POSIX I/O failure
+    ssize_t flush_nx() noexcept;
+    ssize_t flush();
+
+    // write() functions, Used to write many types.
+    //
+    // All calls to write() will be buffered and not necessarily written to the file descriptor except
+    // in these situations:
+    // 1. The size of the passed message is larger than the buffer size.
+    // 2. The size of the passed message is larger than the current buffer free space.
+    // 3. "flush()" (or its exception-free variant: flush_nx()) is called.
+    // 4. A "flush" object is passed to write() functions (equivalent to calling flush()).
+    // 5. An "endl" object is passed to write() functions (equivalent to writing \n and
+    //    calling flush()).
+    // 6. The object is destructed (which calls flush_nx()).
+    //
+    // Multiple arguments can be passed in a single call as well (due to the templated overload), e.g:
+    //
+    //      output_a.write("This is an example message.\n");
+    //      output_b.write("2 + 2 equals: ", 2 + 2, endl);
+    //
+    // RETURN: true or false on failure cases or null/zero-length string.
+    //
+    // NOTE: write() does return true on null/zero-length.
+    //
+    // NOTE: write() does return true on partial write, Use write_buf() and check its return, if you
```

Comment:
`write()` must return false if it cannot write everything that was supposed to be written. This needs to be fixed.